rattan-core/src/cells/token_bucket.rs

```diff
@@ -0,0 +1,1173 @@
+                                        }
+                                    }
+                                }
+                            }
+                            None => {
+                                // channel closed
+                                return None;
+                            }
+                        }
+                    }
+                    _ = self.timer.sleep(self.next_available - Instant::now()) => {
+                        break;
+                    }
+                }
+            }
+
+            // wait for packet to arrive and check it
+            while self.packet_queue.is_empty() {
+                // the queue is empty, wait for packets to arrive
+                tokio::select! {
+                    biased;
+                    Some(config) = self.config_rx.recv() => {
+                        self.set_config(config);
+                    }
+                    recv_packet = self.egress.recv() => {
+                        match recv_packet {
+                            Some(new_packet) => {
+                                match self.state.load(std::sync::atomic::Ordering::Acquire) {
+                                    0 => {
+                                        return None;
+                                    }
+                                    1 => {
+                                        return Some(new_packet);
+                                    }
+                                    _ => {
+                                        if self.token_bucket.buffer == -1 {
+                                            return Some(new_packet);
+                                        }
+                                        // check the size of packet and drop or enter the queue
+                                        let packet_size = new_packet.l3_length() + self.packet_queue.get_extra_length();
+                                        if packet_size > self.max_size as usize {
+                                            // drop the packet
+                                            return None;
+                                        }
+                                        else {
+                                            self.packet_queue.enqueue(new_packet);
+                                        }
+                                    }
+                                }
+                            }
+                            None => {
+                                // channel closed
+                                return None;
+                            }
+                        }
+                    }
+                }
+            }
+            // packet arrived, check the size of packet and number of tokens
+
+            // check if the packet is not larger than max_size
+            if self.packet_queue.get_front_size().unwrap() > self.max_size as usize {
+                return None;
+            }
```

Comment:
Can you please confirm if this check is redundant? It seems there is no possibility of enqueuing such a packet to the queue.

Reply:
When the buffer or rate is modified at runtime, those packets which were enqueued before modification could be larger than `max_size`. So I check the size of the first packet before any dequeueing.

Reply:
In that case, should you enqueue all the packets? As you state, the buffer or rate can be modified at runtime. If the limit on packet size is enlarged, previously dropped packets can be dequeued. Please confirm when is the best time to drop such packets according to source code or some other references, as undropped packets would still take the capacity of buffer, which will later influence the execution process of the tbf.